server/src/main/java/com/cloud/network/element/ConfigDriveNetworkElement.java

```diff
@@ -639,7 +639,10 @@ private boolean deleteConfigDriveIso(final VirtualMachine vm) throws ResourceUna
 
         if (location == Location.SECONDARY) {
             dataStore = _dataStoreMgr.getImageStoreWithFreeCapacity(vm.getDataCenterId());
-            agentId = findAgentIdForImageStore(dataStore);
+            if (!VirtualMachineManager.VmConfigDriveOnPrimaryPool.valueIn(vm.getDataCenterId()) &&
+                    !VirtualMachineManager.VmConfigDriveForceHostCacheUse.valueIn(vm.getDataCenterId()) && dataStore != null) {
+                agentId = findAgentIdForImageStore(dataStore);
```

Comment:
@Pearl1594 will `agentId` be null if the conditions do not match ?

Reply:
yes @weizhouapache it could be null - and it's handled by https://github.com/apache/cloudstack/blob/main/server/src/main/java/com/cloud/network/element/ConfigDriveNetworkElement.java#L651 - however, this exception, doesn't prevent the deletion of the VM